packages/polling-controller/src/PollingController.ts

```diff
@@ -1,150 +1,144 @@
-import { BaseControllerV2 } from '@metamask/base-controller';
-import type {
-  RestrictedControllerMessenger,
-  StateMetadata,
-} from '@metamask/base-controller';
+import { BaseController, BaseControllerV2 } from '@metamask/base-controller';
 import type { NetworkClientId } from '@metamask/network-controller';
-import type { Json } from '@metamask/utils';
 import { v4 as random } from 'uuid';
 
-export type PollingCompleteType<N extends string> = {
-  type: `${N}:pollingComplete`;
-  payload: [string];
-};
+// eslint-disable-next-line @typescript-eslint/ban-types
+type Constructor = new (...args: any[]) => {};
 
 /**
- * PollingController is an abstract class that implements the polling
- * functionality for a controller. It is meant to be extended by a controller
- * that needs to poll for data by networkClientId.
+ * PollingControllerMixin
  *
+ * @param Base - The base class to mix onto.
+ * @returns The mixin.
  */
-export default abstract class PollingController<
-  Name extends string,
-  State extends Record<string, Json>,
-  messenger extends RestrictedControllerMessenger<
-    Name,
-    any,
-    PollingCompleteType<Name> | any,
-    string,
-    string
-  >,
-> extends BaseControllerV2<Name, State, messenger> {
-  readonly #intervalLength: number;
+function PollingControllerMixin<TBase extends Constructor>(Base: TBase) {
+  /**
+   * PollingController is an abstract class that implements the polling
+   * functionality for a controller. It is meant to be extended by a controller
+   * that needs to poll for data by networkClientId.
+   *
+   */
+  abstract class PollingControllerBase extends Base {
+    readonly #networkClientIdTokensMap: Map<NetworkClientId, Set<string>> =
+      new Map();
 
-  private readonly networkClientIdTokensMap: Map<NetworkClientId, Set<string>> =
-    new Map();
+    readonly #intervalIds: Record<NetworkClientId, NodeJS.Timeout> = {};
 
-  private readonly intervalIds: Record<NetworkClientId, NodeJS.Timeout> = {};
+    #callbacks: Map<
+      NetworkClientId,
+      Set<(networkClientId: NetworkClientId) => void>
+    > = new Map();
 
-  constructor({
-    name,
-    state,
-    messenger,
-    metadata,
-    pollingIntervalLength,
-  }: {
-    name: Name;
-    state: State;
-    metadata: StateMetadata<State>;
-    messenger: messenger;
-    pollingIntervalLength: number;
-  }) {
-    super({
-      name,
-      state,
-      messenger,
-      metadata,
-    });
+    #intervalLength = 1000;
 
-    if (!pollingIntervalLength) {
-      throw new Error('pollingIntervalLength required for PollingController');
+    getIntervalLength() {
+      return this.#intervalLength;
     }
 
-    this.#intervalLength = pollingIntervalLength;
-  }
+    setIntervalLength(length: number) {
+      this.#intervalLength = length;
+    }
```

Comment:
does intervalLength need to be private if we're just exposing it?

Reply:
We could make it `protected` so that the extending class could decide to expose it or not. I doubt we need this to be configurable at runtime very often, if ever.

Reply:
Can't make it protected with the mixin pattern, i get this error:

```
Property 'intervalLength' of exported class expression may not be private or protected.ts(4094)
```

![image](https://github.com/MetaMask/core/assets/364566/f2f5b07b-48d7-45b9-863d-105518c71811)

Reply:
I think there is a way to do this, just not if I have an `abstract` class where the class has to be named.

> see: https://github.com/microsoft/TypeScript/issues/30355